File: AdministrationServer/repository/src/main/java/by/delta/repository/IRepository.java
package by.delta.repository;

import by.delta.model.abstractmodel.AbstractModel;
import by.delta.specification.abstractspecification.AbstractCriteriaQuerySpecification;

import java.util.List;

public interface IRepository<Entity extends AbstractModel> {

    Entity save(Entity item);

    Entity update(Entity item);

    void remove(Long id);

    List<Entity> query(AbstractCriteriaQuerySpecification<Entity> specification, int limit, int offset);
}

File: AdministrationServer/repository/src/main/java/by/delta/repository/IRepository.kt
package by.delta.repository

import by.delta.model.abstractmodel.AbstractModel
import by.delta.specification.abstractspecification.AbstractCriteriaQuerySpecification
import by.delta.specification.abstractspecification.ICountSpecification



interface IRepository<Entity : AbstractModel> {

    fun save(item: Entity): Entity

    fun update(item: Entity): Entity

    fun remove(id: Long?)

    fun query(specification: AbstractCriteriaQuerySpecification<Entity>, limit: Int, offset: Int): List<Entity>

    fun countQuery(specification: ICountSpecification): Long?
}